keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,558 @@
+
+2. Relax API validation on PVC update so that updating ResourceProvisionedIOPS and ResourceProvisionedThroughput pvc.Spec.Resoures is allowed, as long as requested size is within the range specified by the cluster admin.
+
+3. Add new statuses in PVC API called - pvc.Status.AllocatedResourceStatus for the purpose of tracking status of ProvisionedIOPS and ProvisionedThroughput change. Existing fields can be found [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L520-L534). The new statuses will be added are:
+
+```go
+type ClaimResourceStatus string
+const (
+    PersistentVolumeClaimControllerUpdateIopsProgress ClaimResourceStatus = "ControllerUpdateIopsInProgress"
+    PersistentVolumeClaimControllerUpdateIopsPending ClaimResourceStatus = "ControllerUpdateIopsPending"
+    PersistentVolumeClaimControllerUpdateIopsFailed ClaimResourceStatus = "ControllerUpdateIopsFailed"
+    PersistentVolumeClaimControllerUpdateThroughputInProgress ClaimResourceStatus = "ControllerUpdateThroughputInProgress"
+    PersistentVolumeClaimControllerUpdateThroughputPending ClaimResourceStatus = "ControllerUpdateThroughputPending"
+    PersistentVolumeClaimControllerUpdateThroughputFailed ClaimResourceStatus = "ControllerUpdateThroughputFailed"
+)
+// PersistentVolumeClaimStatus represents the status of PV claim
+type PersistentVolumeClaimStatus struct {
+    ...
+    // AllocatedResourceStatus stores a map of resource that is being expanded
+    // and possible status that the resource exists in.
+    // Some examples may be:
+    //  pvc.status.allocatedResourceStatus["storage"] = "ControllerResizeInProgress"
+    //  pvc.status.allocatedResourceStatus["iops"] = "ControllerUpdateIopsInProgress/ControllerUpdateIopsPending/ControllerUpdateIopsFailed"
+    //  pvc.status.allocatedResourceStatus["throughput"] = "ControllerUpdateThroughputInProgress/ControllerUpdateThroughputPending/ControllerUpdateThroughputFailed"
+    AllocatedResourceStatus map[ResourceName]ClaimResourceStatus
+}
+```
+
+
+
+4. Add new CSI API ControllerUpdateIOProvisioning, when there is a change of iops/throughput in PVC, external-resizer triggers a ControllerUpdateIOProvisioning operations against a CSI endpoint.
+
+5. Update quota code to use min and max(pvc.Spec.Resources, pvc.Status.AllocatedResources) for ProvisionedIOPS and ProvisionedThroughput when evaluating usage for PVC.
+
+![Update Provisioned IO Flow](./qos_flow.png)
+
+Whenever resize controller or kubelet modifies pvc.Status (such as when setting both AllocatedResources and AllocatedResourceStatus) - it is expected that all changes to pvc.Status are submitted as part of same patch request to avoid race conditions.
+
+IO provisioning parameters can be considered as best-effort parameters, we should report the status of bad parameters and the volume would fall back to a workable default configuration.
```

Comment:
Need clarification on what ‘best-effort’ means. What’s the expected behavior in case of illegal configurations? What’s the expected behavior in case a driver is able to handle of subset of requested QoS changes?

Reply:
Define illegal configurations?